File: parcel-locker-cpp/library/src/managers/DeliveryManager.cpp
//
// Created by student on 09.06.2021.
//

#include "managers/DeliveryManager.h"
#include "model/Delivery.h"
#include "model/Locker.h"
#include "model/Client.h"
#include  <boost/uuid/random_generator.hpp>
#include "exceptions/DeliveryManagerException.h"

DeliveryManager::DeliveryManager() {
    currentDeliveries = std::make_shared<DeliveryRepository>();
    archivedDeliveries = std::make_shared<DeliveryRepository>();
}

DeliveryPtr DeliveryManager::makeDelivery(const float &basePrice, const float &width, const float &height, const float &length,
                              const float &weight,
                              bool isFragile, const ClientPtr &shipper, const ClientPtr &receiver,
                              const LockerPtr &locker) {
    boost::uuids::random_generator gen;
    boost::uuids::uuid id = gen();
    DeliveryPtr delivery = std::make_shared<Delivery>(basePrice, id, width, height, length, weight, isFragile, shipper, receiver, locker);
    currentDeliveries->add(delivery);
    return delivery;
}

DeliveryPtr DeliveryManager::makeDelivery(const float &basePrice, bool priority, const ClientPtr &shipper,
                                          const ClientPtr &receiver, const LockerPtr &locker) {
    boost::uuids::random_generator gen;
    boost::uuids::uuid id = gen();
    DeliveryPtr delivery = std::make_shared<Delivery>(basePrice, id, priority, shipper, receiver, locker);
    currentDeliveries->add(delivery);
    return delivery;
}

void DeliveryManager::putInLocker(const DeliveryPtr &delivery, const std::string &accessCode) {
    delivery->getLocker()->putIn(delivery->getId(), delivery->getReceiver()->getTelNumber(), accessCode);
    delivery->setStatus(Ready_for_pickup);
    ///wkładamy przesyłkę           szukamy pustego                         podajemy numer odbiorcy          kod odbioru
    ///do ustalonego lockera
}

bool
DeliveryManager::takeOutDelivery(const LockerPtr &locker, const ClientPtr &receiver, const std::string &accessCode) {
    boost::uuids::uuid deliveryId = locker->takeOut(receiver->getTelNumber(), accessCode);
    DeliveryPtr delivery = currentDeliveries->findById(deliveryId);
    if(!delivery)
        return false;
    currentDeliveries->remove(delivery);
    archivedDeliveries->add(delivery);
    delivery->setStatus(Received);
    return true;
}

std::vector<DeliveryPtr> DeliveryManager::getAllClientDeliveries(const ClientPtr &client) {
    try {
        if (!client)
            throw DeliveryManagerException("client is a nullptr!");
    }
    catch (DeliveryManagerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }
    return currentDeliveries->findBy([&client](const DeliveryPtr& delivery){ return delivery->getReceiver()->getTelNumber() == client->getTelNumber(); });
}

float DeliveryManager::checkClientShipmentBalance(const ClientPtr &client) {
    float balance = 0;
    try {
        if(!client)
            throw DeliveryManagerException("client is a nullptr!");
        for (const auto &delivery : currentDeliveries->findAll()) {
            if (delivery->getShipper() == client)
                balance += delivery->getCost();
        }
    }

    catch (DeliveryManagerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }

    return balance;
}

const DeliveryRepositoryPtr &DeliveryManager::getArchivedDeliveries() const {
    return archivedDeliveries;
}

std::vector<DeliveryPtr> DeliveryManager::getAllReceivedClientDeliveries(const ClientPtr &client) {
    try {
        if (!client)
            throw DeliveryManagerException("client is a nullptr!");
    }
    catch (DeliveryManagerException &e) {
        std::cerr << e.what() << std::endl;
        throw;
    }
    return archivedDeliveries->findBy([&client](const DeliveryPtr& delivery){ return delivery->getReceiver()->getTelNumber() == client->getTelNumber(); });

}

File: parcel-locker/src/main/java/pl/pas/parcellocker/managers/DeliveryManager.java
package pl.pas.parcellocker.managers;

import pl.pas.parcellocker.exceptions.DeliveryManagerException;
import pl.pas.parcellocker.model.Client;
import pl.pas.parcellocker.model.Delivery;
import pl.pas.parcellocker.model.Locker;
import pl.pas.parcellocker.repositories.DeliveryMongoRepository;
import pl.pas.parcellocker.repositories.kafka.ProducerHandlerImpl;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

import static pl.pas.parcellocker.model.DeliveryStatus.READY_TO_PICKUP;
import static pl.pas.parcellocker.model.DeliveryStatus.RECEIVED;

public class DeliveryManager {

    private final DeliveryMongoRepository deliveries;

    public DeliveryManager() {
        deliveries = new DeliveryMongoRepository(new ProducerHandlerImpl());
    }

    public Delivery makeParcelDelivery(
        BigDecimal basePrice,
        double width,
        double height,
        double length,
        double weight,
        boolean isFragile,
        Client shipper,
        Client receiver,
        Locker locker
    ) {
        Delivery delivery = new Delivery(basePrice, width, height, length, weight, isFragile, shipper, receiver, locker);
        deliveries.add(delivery);
        return delivery;
    }

    public Delivery makeListDelivery(
        BigDecimal basePrice,
        boolean isPriority,
        Client shipper,
        Client receiver,
        Locker locker
    ) {
        Delivery delivery = new Delivery(basePrice, isPriority, shipper, receiver, locker);
        deliveries.add(delivery);
        return delivery;
    }

    public void putInLocker(Delivery delivery, String accessCode) {
        Locker locker = delivery.getLocker();
        locker.putIn(delivery, delivery.getReceiver().getTelNumber(), accessCode);
        delivery.setStatus(READY_TO_PICKUP);
        deliveries.update(delivery);
    }

    public boolean takeOutDelivery(Locker locker, Client receiver, String accessCode) {
        UUID deliveryId = locker.takeOut(receiver.getTelNumber(), accessCode);

        if (deliveryId != null) {
            Delivery delivery = deliveries.findById(deliveryId);
            delivery.setStatus(RECEIVED);
            deliveries.update(delivery);
            return true;
        }
        return false;
    }

    public BigDecimal checkClientShipmentBalance(Client client) {
        BigDecimal balance = BigDecimal.ZERO;
        if (client == null)
            throw new DeliveryManagerException("client is a nullptr!");
        for (Delivery delivery : deliveries.findAll()) {
            if (delivery.getShipper().equals(client))
                balance = balance.add(delivery.getCost());
        }

        return balance;
    }

    public List<Delivery> getAllClientDeliveries(Client client) {
        return deliveries.findByClient(client);
    }

    public List<Delivery> getAllReceivedClientDeliveries(Client client) {
        return deliveries.findArchivedByClient(client);
    }
}
